### PR TITLE
Profiles feature branch in Engine

### PR DESCRIPTION
**Pull request by mattias-p**
This is a work in progress for #383.

### CODE REVIEW

**Comment by sandoche2k** [comment]
@vlevigneron can you please review

**Comment by matsduf** [comment]
My picture is that this is not ready to be merged now. This PR is here to see what changes it will introduce, isn't it, @mattias-p?

**Comment by vlevigneron** [comment]
Should we, anyway, push the corresponding code in this branch ?
I mean, do you plan to have this branch dedicated to documentation only
or should we do all profiles stuff in it.

By the way, do we plan to provide backward compatibility with current
configuration methods ?

le 25 sept., matsduf a ?crit :
> My picture is that this is not ready to be merged now. This PR is here to see what changes it will introduce, isn't it, @mattias-p?
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/zonemaster/zonemaster-engine/pull/386#issuecomment-424255847

-- 
	Vincent Levigneron  A.F.N.I.C.  Vincent.Levigneron@afnic.fr

**Comment by matsduf** [comment]
Good questions. Let's discuss in meeting.

**Comment by sandoche2k** [comment]
If we get one more approval, this PR will be approved

File: lib/Zonemaster/Engine/Profile.pm [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+The values of the second level hashes are mapped to severity levels.
+The default severity level is C<DEBUG> for tags not found in the
+C<test_levels> item.
+
+N.B. At the time of writing there is no authoritative documentation for
+the default value of this property.
+For the time being, please refer to the file
+located by L<dist_file("Zonemaster-Engine",
+"default.profile")|File::ShareDir/dist_file>.
+
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Profile.pm]: The updated test cases have default values. Please refer to those.

File: lib/Zonemaster/Engine/Profile.pm [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+valid values.
+
+Default values are listed here as specified in the distributed default
+profile JSON file.
+
+=head2 resolver.defaults.usevc
+
+A boolean. If C<true>, only use TCP. Default C<false>.
+
+=head2 resolver.defaults.retrans
```
[the remark this answers is not in the document]
Comment (mattias-p) [reply lib/Zonemaster/Engine/Profile.pm]: Yes, in two places:
* https://github.com/zonemaster/zonemaster-engine/blob/profiles/lib/Zonemaster/Engine/Nameserver.pm#L81
* https://github.com/zonemaster/zonemaster-engine/blob/profiles/lib/Zonemaster/Engine/Nameserver.pm#L282

File: lib/Zonemaster/Engine/Profile.pm [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+
+A boolean. If C<true>, sets the DO flag in queries. Default C<false>.
+
+=head2 resolver.defaults.recurse
+
+A boolean. If C<true>, sets the RD flag in queries. Default C<false>.
+
+This should almost certainly be kept C<false>.
+
+=head2 resolver.defaults.retry
```
[the remark this answers is not in the document]
Comment (mattias-p) [reply lib/Zonemaster/Engine/Profile.pm]: Yes, in two places:
* https://github.com/zonemaster/zonemaster-engine/blob/profiles/lib/Zonemaster/Engine/Nameserver.pm#L81
* https://github.com/zonemaster/zonemaster-engine/blob/profiles/lib/Zonemaster/Engine/Nameserver.pm#L282

File: lib/Zonemaster/Engine/Profile.pm [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+
+A non-negative integer. The number of times a query is sent before we
+give up. Default 2.
+
+If set to zero, no queries will be sent at all, which isn't very useful.
+
+=head2 resolver.defaults.igntc
```
[the remark this answers is not in the document]
Comment (mattias-p) [reply lib/Zonemaster/Engine/Profile.pm]: Yes, in two places:
* https://github.com/zonemaster/zonemaster-engine/blob/profiles/lib/Zonemaster/Engine/Nameserver.pm#L81
* https://github.com/zonemaster/zonemaster-engine/blob/profiles/lib/Zonemaster/Engine/Nameserver.pm#L282

File: docs/Profiles.md [hunk 5]
```diff
@@ -0,0 +1,17 @@
+# Profiles
+
+## Default profile
+
+The default profile is documented in the [profile properties] section
+of the Zonemaster::Engine::Profile module.
+
+At the time of writing there is no authoritative documentation for the
+default value of the `test_levels` property.
```
[the remark this answers is not in the document]
Comment (mattias-p) [reply docs/Profiles.md]: Do you have a link I can use?

File: lib/Zonemaster/Engine/Profile.pm (hunk 1 above)
Comment (mattias-p) [reply lib/Zonemaster/Engine/Profile.pm]: Addressing this in #441.

File: lib/Zonemaster/Engine/Profile.pm [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+=head1 INSTANCE METHODS
+
+=head2 get
+
+Get the value of a property.
+
+    my $value = $profile1->get( 'net.ipv6' );
+
+Returns value of the given property, or C<undef> if the property is unset.
+The returned value is a deep copy.
```
[the remark this answers is not in the document]
Comment (mattias-p) [reply lib/Zonemaster/Engine/Profile.pm]: Addressing this in #441.

File: lib/Zonemaster/Engine/Profile.pm (hunk 1 above)
Comment (mattias-p) [reply lib/Zonemaster/Engine/Profile.pm]: Addressing this in #441.

File: docs/Profiles.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (matsduf) [reply docs/Profiles.md]: OK with #441

**Comment by matsduf** [comment]
We have already approved the parts so I have no problem to approve, but I think Travis should work first.

**Comment by vlevigneron** [comment]
The problem with Travis is that it does not use the last modified version but the released one (I guess). The fallback method exists only in https://github.com/zonemaster/zonemaster-ldns/tree/develop.

**Pull request merged**